opencog/rule-engine/forwardchainer/ForwardChainer.cc

```diff
@@ -144,19 +147,50 @@ void ForwardChainer::do_chain(ForwardChainerCallBack& fcb,
      init_sources = _as.get_outgoing(hsource);
     else
         init_sources.push_back(hsource);
+
     _fcmem.update_potential_sources(init_sources);
 
-    _fcmem.set_source(fcb.choose_next_source(_fcmem)); //set initial source
     auto max_iter = _configReader.get_maximum_iterations();
 
     while (_iteration < max_iter /*OR other termination criteria*/) {
+
         _log->info("Iteration %d", _iteration);
 
-        do_step(fcb);
+        if(rule_weight.empty())
+        {
+            //! Choose next source.
+            _log->info("[ForwardChainer] setting next source");
+
+            _fcmem.set_source(choose_next_source(_fcmem));
+
+            // Choose matching rules whose input matches with the source.
+            vector<Rule*> matched_rules = choose_rules(_fcmem);
```

Comment:
Got it, so we need to have 2 different rule selection/application methods (1) tournament selection (of a single rule) and (2) iterate through all rules.

Let's add this to the google spec doc so we're all on the same page.

@misgeatgit, @williampma , @ngeiswei should we discuss how to best implement for the different use cases? Perhaps a PLNForwardChainer class and a R2LForwardChainer class that inherit from ForwardChainer? I think the 2 rule selection/application methods (single selection vs all rules) should be implemented separately from the use cases (PLN vs R2L), because the methods should be available to other future use cases.